### PR TITLE
[WIP] pull-multiplex - pull-streams all the way down

### LINKED ISSUE #14: Migrate to pull-streams
To be compatible with the rest of the libp2p, this needs to be refactored to support pull-streams interface. 

Ref: https://github.com/libp2p/interface-stream-muxer/pull/7

### PR DESCRIPTION
**Pull request by dignifiedquire**
Current State

- [ ] New multiplex implementation using only pull-streams in `src/multiplex.js`
  - [ ] Spec compliant
    - [ ] Do not encode the header with a length prefix
    - [ ] Check maximum size of data 
  - [ ] finish `destroy` method
  - [ ] All tests passing
    - [ ] Mega stress test, does not run out of memory, but doesn't finish either.
    - [ ] Close tests
  - [ ] Split into their own package 
    - [ ]  `src/pull-end`
    - [ ] `src/pull-split`
    - [ ] `src/multiplex`
- [ ] Cleanup code
- [ ] https://github.com/libp2p/interface-stream-muxer/compare/fix/multiplex?expand=1
- [ ] Add documentation

### CODE REVIEW

**Comment by diasdavid** [comment]
Have you tested this with the stress tests? Any luck?

**Comment by dignifiedquire** [comment]
@diasdavid 

- stress tests from interface-stream-muxer pass except on error case
- https://github.com/libp2p/js-libp2p-spdy/issues/48#issuecomment-269137170 is the interesting part

File: test/promise.spec.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/* eslint-env mocha */
```
Comment (dignifiedquire) [inline test/promise.spec.js]: This is super important :P no I just forgot to delete it before comitting, fell free to remove

**Comment by VictorBjelkholm** [comment]
There is an updated branch based on this work but by using `multiplex` from npm rather than a re-implementation of the core of multiplex here: https://github.com/libp2p/js-libp2p-multiplex/pull/49

**Comment by VictorBjelkholm** [comment]
`pull-wrap` branch was merged in 0f3bc8ce0ca9b62a81e425866690f1d9d85a96c2 and exposes a pull-stream interface identical to the one here, but with using `multiplex` module instead of our implementation.

We'll keep this one open for the future.

**Comment by diasdavid** [comment]
@dryajov now that you are attempting #73, make sure to go through the work @dignifiedquire did here

**Pull request closed without merging**